### DISCUSSION TITLE
[1.2.0]지금 멀티서버내 이동 및 애니메이션 동기화 관해서 오류가 많습니다.

### DISCUSSION THREAD

**Discussion by ted002** [naverz/zepeto-studio-kor#645] [Issue Report (Legacy)]
현재 제페토측에서 제공하는 멀티서버의 애니메이션 동기화는 MoveToPosition을 활용한 레퍼런스 뿐입니다.

허나 MoveToPosition을 활용한 이동의 경우.
-> 짧은 딜레이가 생기지만, 자체적인 이동함수이기 때문에 뛰기 애니메이션에 한해서 문제 없음.
-> 단, 걷거나 가볍게 뛰는 경우 다른 기기에서는 뛰는 애니메이션이 계속 뚝뚝 끊기면서 재생되고, 이동속도가 무조건 최대속도라서 가다서다를 반복한다.
->State를 통해 점프 상태를 파악하기에, 내 기기에서 점프를 1회 하고 끝났을에도 다른 기기에서 점프상태가 끝나지 않아 한번 더 점프하는 경우가 많음.

그렇다고 받아온 포지션을 강제로 적용하는 경우에는 
->짧은 딜레이가 있음, 자연스러운 이동이 되지 않고 끊기는 느낌이 있음
->애니메이션 동기화를 위해 State를 통해 확인하려고 해도 기본에 쓰던 walk, run, JumpIdle, JumpMove등의 State가 나타나지 않기 때문에 걷던 뛰던 무조건 뛰는 State, 102번 State밖에 나오지 않음. 구분할수가 없음.

현재 모션V2에서는 걷기, 가볍게 뛰기, 전력질주, 일반점프, 뛰면서 점프, 회전, 낙하등의 애니메이션이 있습니다. 허나 MoveToPosition을 활용해서 위치 동기화 및 이동 애니메이션 동기화를 할 경우, 걷기, 가볍게 뛰기, 전력질주의 구분 및 이동속도 구분은 사라지고 오직 전력질주 이동속도 + 전력질주 애니메이션만 나오기 때문에 오류가 생깁니다.

그렇다고 레퍼런스가 아닌 제 나름대로 위치 동기화 및 애니메이션 동기화를 하고싶어도, State 자체가 100번 이후의 state만 사용하고 구분이 제대로 되어있지 않기에 결국 동일한 문제가 생깁니다.

물론 사적으로 제가 만든 액션, 혹은 자세등을 동기화 하는것은 어렵지 않습니다. 제가 말씀드리고 싶은건, 멀티서버 내 위치 및 이동 애니메이션 동기화의 레퍼런스로 MoveToPosition을 알려주셨다면, 적어도 해당 함수가 보통의 이동, 회전이나 걷기, 빨리걷기등의 구분정도는 할 수 있어야 한다고 생각합니다. 아니면 State가 제대로 구분되어 나오던지요.

긴 글 읽어주셔서 감사드리며, 개선 꼭 좀 부탁드립니다.

**Comment by smilejsu82** [comment]
월드 1.2.0 입니다 
모션v2 사용중입니다 

저도 그부분은 많이 고생했는데요 

완전히 안되는건 아니드라구요 

저 같은 경우 달리면서 과일 버튼을 누르면 근처 타겟을 찾고 사거리에 있으면 달리면서 바라봐야 했습니다 
회전동기화가 매우 까다로웠는데요 결국 LateUpdate에서 해결 했습니다 

그리고 Dash버튼을 누르면 달리면서 굴렀다가 다시 달려야 했는데요 이부분은 애니메이션은 레이어를 하나 더 만들어서 동기화 했습니다 

덕분에 코드가 많이 복잡해지기는 했습니다만 정상동작 하는것을 확인 했습니다 ^^

https://user-images.githubusercontent.com/50188795/164382599-8a5cbe78-fff0-4079-8d88-e3591c258951.mp4

**Reply by ted002** [reply]
음...말씀해주신 것 만 가지고는 잘 이해가 안가네요...
1. LateUpdate를 어떻게 활용하셨는지 잘 이해가 안갑니다.
LateUpdate라고 하시는걸 보면 ClientStarter에 코드에 LateUpdate를 넣고 활용하신거라고 말씀하신거로 이해했는데, 그럼 해당 클라이언트 스타터에 플레이어와 세션 아이디를 저장해서 사용하는건지, 어떤방식으로 회전동기화를 하셨는지 잘 이해가 안갑니다.

2. 레이어의 활용을 어떻게 하셨는지 이해가 안갑니다
레이어에 대한 이해 자체가 부족해서 그런지, 잘 이해가 안갑니다. 신규 레이어를 추가해서 사용한 것인가요? 아니면 애니메이션 자체를 새로 만드신건가요?

4. 기본 애니메이션들(걷기, 뛰기,전력질주)는 어떻게 구분하셨는지, 위치동기화는 어떤 방법을 사용하셨는지 궁금합니다!

**Reply by smilejsu82** [reply]
1. 저같은경우 PlayerController스크립트(사용자 정의)를 플레이어가 생성될때 붙여 주었습니다 

자신이 움직이고 있는 케릭터가 과일 버튼을 눌렀을 경우 타겟을 찾고 먼저 스스로 회전 합니다 
필요한 데이터를 담아 서버로 전송 해요 
(이때 전 누가 누구를 바라봐야 하는지에 대한 정보를 담았습니다)
`playerController.lookAtLateAction = (req: req_lookat) => {
                    this.SendLookAtLate(req);
                };`

전송 할때는 다음과 같이 json문자열로 전송 했습니다 
`    private SendLookAtLate(req) {
        var json = JSON.stringify(req);
        console.log(json);
        const data = new RoomData();
        data.Add("req_lookat", json);
        this.room.Send("onLookAtLate", data.GetObject());
    }
`
서버 코드 입니다 
그냥 받은 데이터를 그대로 broadcast 했어요 
`this.onMessage("onLookAtLate", (client, message) => {
      this.broadcast("onLookAtLate", message.req_lookat);
    });`

그리고 클라이언트에서 받아 해당 플레이어를 찾고 타겟을 넣어줬습니다 
`var targetPlayer = ClientStarter.players.get(req.toHash);
                            playerController.SetLookAtTarget(targetPlayer.gameObject);
                            playerController.ThrowFruit();`


다음과 같이 스크립트를 작성해서 회전 시켜 주었습니다 
` LateUpdate() {

        if (this.lateLookAtTargetGo != null) {
            var dir = (this.lateLookAtTargetGo.transform.position - this.transform.position).normalized;
            var angle = Mathf.Atan2(dir.x, dir.z) * Mathf.Rad2Deg;
            this.transform.rotation = Quaternion.AngleAxis(angle, UnityEngine.Vector3.up);
        }
    }`


2. ZepetoAnimatorV2를 가져와 다음과 같이 수정 했습니다 
![image](https://user-images.githubusercontent.com/50188795/164388707-1b6fa770-f327-43c9-9792-69b30d91aa4c.png)
![image](https://user-images.githubusercontent.com/50188795/164388864-71d994c1-0813-4cb9-abad-10dce2c39dec.png)


3. 기본 애니메이션들은 안건드렸습니다 위치동기화는 기본 코드를 사용했습니다 
`// [RoomState] player 인스턴스의 state가 갱신될 때마다 호출됩니다.
                    player.OnChange += (changeValues) =>
                        this.OnUpdatePlayer(sessionId, player);`

`private OnUpdatePlayer(sessionId: string, player: Player) {

        const position = this.ParseVector3(player.transform.position);

        const zepetoPlayer = ZepetoPlayers.instance.GetPlayer(sessionId);
        zepetoPlayer.character.MoveToPosition(position);

        if (player.state === CharacterState.JumpIdle || player.state === CharacterState.JumpMove)
            zepetoPlayer.character.Jump();
    }`


말로 설명하기 난해한 부분이 있습니다 참고만 하세요 ~
이게 정답이 아닐수도 있어요 :) 
그럼 화이팅 👍

**Reply by ted002** [reply]
감사합니다 한번 해볼게요!

**Reply by smilejsu82** [reply]
> 감사합니다 한번 해볼게요!

어떤걸 구현 하려고 하세요? 혹시 설명해주시면 예제를 한번 만들어 드릴게요 😀

**Reply by ted002** [reply]
정확히 이걸 구현하겠다기 보다는, 지금 제페토 내에서 이번에 V2 모션 생기면서 애니메이션이 여러가지가 생겼는데, 막상 멀티서버를 만들어서 해당 레퍼런스대로 구현을 해보니 기껏 만들어진 V2의 모션들이 살아나지 못하는거 같아서요. 개인적으로 여러가지 방법을 써봤는데 잘 안되던 참에 애니메이터의 layer 활용법이나 지금 알려주신 LateUpdate부분들도 한번 활용해서 여러모로 시험해보려고 합니다!

**Reply by smilejsu82** [reply]
> 정확히 이걸 구현하겠다기 보다는, 지금 제페토 내에서 이번에 V2 모션 생기면서 애니메이션이 여러가지가 생겼는데, 막상 멀티서버를 만들어서 해당 레퍼런스대로 구현을 해보니 기껏 만들어진 V2의 모션들이 살아나지 못하는거 같아서요. 개인적으로 여러가지 방법을 써봤는데 잘 안되던 참에 애니메이터의 layer 활용법이나 지금 알려주신 LateUpdate부분들도 한번 활용해서 여러모로 시험해보려고 합니다!

그렇군요 ~ 여러가지들을 테스트 해보시고 공유 해주시면 좋을거 같아요 ~ 화이팅 😄

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, V2 전용 모션에 대한 세부 State의 경우, 다음과 같이 접근하여 확인 할 수 있습니다.


enum CharacterState { Invalid = 0, Idle = 1, Walk = 2, Run = 3, JumpIdle = 4, JumpMove = 5, Teleport = 20, Gesture = 30, Move = 102, MoveTurn = 103, Jump = 104, Stamp = 106, Falling = 108, Landing = 109 }

ZepetoCharacter.MotionV2.CurrentLandingState;
ZepetoCharacter.MotionV2.CurrentMoveState;
ZepetoCharacter.MotionV2.CurrentJumpState;

enum CharacterMoveState { None = -1, MoveWalk = 0, MoveRun = 1 }

enum CharacterJumpState { None = -1, JumpIdle = 0, JumpMove = 1, JumpDash = 2, JumpDouble = 3 }

enum CharacterLandingState { None = -1, LandingSlight = 0, LandingDeep = 1, LandingRoll = 2 }


각각 대응하는 기본 State(V2) 에 대해, SubState로서  모션상태를 나타냅니다.
감사합니다.
